packages/flutter_tools/lib/src/plugins.dart

```diff
@@ -351,21 +395,165 @@ Future<List<Plugin>> findPlugins(FlutterProject project, { bool throwOnError = t
     project.directory.path,
     '.packages',
   );
+  final String pubspec = globals.fs.path.join(
+    project.directory.path,
+    'pubspec.yaml',
+  );
+  YamlMap appDependencies;
+  try {
+    final File pubspecFile = globals.fs.file(pubspec);
+    if (pubspecFile.existsSync()) {
+      final dynamic pubspec = loadYaml(pubspecFile.readAsStringSync());
+      appDependencies = pubspec != null ? pubspec['dependencies'] as YamlMap : null;
+    }
+  } on YamlException catch (err) {
+    if (throwOnError) {
+      throwToolExit('Failed to parse pubspec.yaml $err');
+    }
+    globals.printTrace('Failed to parse pubspec.yaml $err');
+  }
   final PackageConfig packageConfig = await loadPackageConfigWithLogging(
     globals.fs.file(packagesFile),
     logger: globals.logger,
     throwOnError: throwOnError,
   );
   for (final Package package in packageConfig.packages) {
     final Uri packageRoot = package.packageUriRoot.resolve('..');
-    final Plugin plugin = _pluginFromPackage(package.name, packageRoot);
+    final Plugin plugin = _pluginFromPackage(
+      package.name,
+      packageRoot,
+      <String>[...appDependencies.keys.cast<String>()],
+    );
     if (plugin != null) {
       plugins.add(plugin);
     }
   }
   return plugins;
 }
 
+/// Metadata associated with the resolution of a platform interface of a plugin.
+class PluginInterfaceResolution {
+  PluginInterfaceResolution({
+    @required this.plugin,
+    this.platform,
+    this.dartClass,
+  }) : assert(plugin != null);
+
+  /// The plugin.
+  final Plugin plugin;
+  // The name of the platform that this plugin implements.
+  final String platform;
+  // The name of the Dart class.
+  final String dartClass;
+
+  Map<String, String> toMap() {
+    return <String, String> {
+      'pluginName': plugin.name,
+      'dartClass': dartClass,
+      'platform': platform,
+    };
+  }
+}
+
+/// Resolves the platform implementation for Dart-only plugins.
+///
+///   * If there are multiple direct pub dependencies on packages that implement the
+///     frontend plugin for the current platform, fail.
+///   * If there is a single direct dependency on a package that implements the
+///     frontend plugin for the target platform, this package is the selected implementation.
+///   * If there is no direct dependency on a package that implements the frontend
+///     plugin for the target platform, and the frontend plugin has a default implementation
+///     for the target platform the default implementation is selected.
+///   * Else fail.
+///
+///  For more details, https://flutter.dev/go/federated-plugins.
+List<PluginInterfaceResolution> resolvePlatformImplementation(
+  List<Plugin> plugins, {
```

Comment:
I don't recall if plugins are sorted by this point, but since the order matters for the errors when their are conflicts, you should ensure they are sorted so the error messages stay consistent.

Reply:
The order of the plugins doesn't affect the entries in the output list. The order of the entries would differ, but that isn't a problem.